File: app/src/main/java/com/seasonfif/learnkotlin/concurrent/FutureTaskDemo.java
package com.seasonfif.learnkotlin.concurrent;

import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.FutureTask;

public class FutureTaskDemo {

    public static void main(String[] args){

        ExecutorService pool = Executors.newFixedThreadPool(3);

        Future f = pool.submit(new Runnable() {
            @Override
            public void run() {

            }
        });

        try {
            Object i = f.get();
            System.out.println(i);
        } catch (InterruptedException e) {
            e.printStackTrace();
        } catch (ExecutionException e) {
            e.printStackTrace();
        }

        Future<Integer> ff = pool.submit(new Callable<Integer>() {
            @Override
            public Integer call() {
                return 3;
            }
        });

        try {
            System.out.println(ff.get());
        } catch (InterruptedException e) {
            e.printStackTrace();
        } catch (ExecutionException e) {
            e.printStackTrace();
        }

        System.out.println("finish");

        FutureTask<String> fff = new FutureTask<>(new Callable<String>() {
            @Override
            public String call() throws Exception {
                return "FutureTask<String>";
            }
        });

        new Thread(fff).start();
        try {
            System.out.println(fff.get());
        } catch (InterruptedException e) {
            e.printStackTrace();
        } catch (ExecutionException e) {
            e.printStackTrace();
        }
        System.out.println("finish");
    }
}

File: app/src/main/java/com/seasonfif/learnkotlin/concurrent/FutureTaskDemo.kt
package com.seasonfif.learnkotlin.concurrent

import java.util.concurrent.Callable
import java.util.concurrent.FutureTask

fun main(args: Array<String>) {

    var future : FutureTask<Int> = FutureTask(CalculateTask())
    var t : Thread = Thread(future)
    t.start()

    //Thread的join和FutureTask的get方法调用均会阻塞主线程的执行
    //然而后者可以同步等待执行的结果
    t.join()

//    println("取得结果：" + future.get())
    println("运行结束")


}

class CalculateTask : Callable<Int>{
    /**
     * Computes a result, or throws an exception if unable to do so.
     *
     * @return computed result
     * @throws Exception if unable to compute a result
     */
    override fun call(): Int {
        var sum = 0
        for (i in 1..1000){
            sum += i
        }
        return sum
    }
}